Hello angular motoko (#277)
dfinity/examples: motoko/hello-angular-motoko/README.md

# hello-angular-motoko
This is an example of a Motoko project with an Angular 14 frontend.

<a name="prerequesites"></a> 
## Prerequesites
- [Install DFX](https://internetcomputer.org/docs/current/developer-docs/quickstart/local-quickstart).
- [Install Angular Globally](https://angular.io/guide/setup-local)
- Install npm packages from the project root:
    ```
    npm install
    ```

> - A https://nns.ic0.app wallet, logged in.
> - ~5 US dollars worth of ICP transfered to your wallet address.
> - As taken from [this guide](https://kyle-peacock.com/blog/dfinity/your-first-canister): 
>  
>     1. Login to your https://nns.ic0.app, select your account and verify your balance.
>     2. Select the [CANISTERS](https://nns.ic0.app/#/canisters) view on the left side navigation menu. 
>     3. Click the blue "Create or Link Canister" button at the bottom of your page to open up the dialog.
>     4. Click the "Create New Canister" button. Then, select the ICP account that you want to use to fund your new canister.
>     5. You'll then be prompted to enter an amount. For now, put at least 2T Cycles.
>     6. Click "Review Cycles Purchase" and Confirm.
>         * You will need to create 2 unique Canisters for this project. Repeat steps 3-6 to create a second canister. 
>     7. You'll need to get the principal from your computer by running from the project root
>         ```
>         dfx identity get-principal
>         ```
>     8. Copy that principal id and head back to your CANISTERS view with your newly created Canisters. Click the blue "Change Controllers" button to open up the controller management UI.
>         * Repeat for the second Canister.
>     9. Enter your principal in the empty second input, and then click "Perform Controller Change". This step authorizes your dfx on your computer to deploy to your newly created Canister. 
>         * Repeat for the second Canister as well.
>     10. Modify the `canister_id.json` file in the project root. Update both "motoko" and "www" "ic" properties.
> 	 
>         ```bash
>         # canister_id.json
>         {
>         "motoko": {
>             # replace this with your unique cansiter from the above guide
>             "ic":"your-unique-canister-goes-here"
>         },
>         "www": {
>             # replace this with a second unique cansiter from the above guide
>             "ic": "another-unique-canister-goes-here"
>         }
>     
>         ```

Reviewer: ```suggestion
- For live IC deployment: A wallet set up as described in the [quickstart guide](https://internetcomputer.org/docs/current/developer-docs/quickstart/hello10mins/).
```
With a configured wallet (which I think is a reasonable demand), this could all be replaced by `dfx canister --network ic create --all --with-cycles 1000000000000`, which `dfx deploy --network ic` automatically does when you also add `--with-cycles 1000000000000`.

Author: @sesi200 
- How does the project know what canisters to deploy to? This is the recommended method as given by @krpeacock. I imagine it will do well for now.
- As you stated before, simply providing a link without explicitly describing the instructions is bad practice.

Reviewer: I didn't explain myself enough the first time around, sorry. The reasons I didn't like the link to Kyle's article are:
- It's not the standard way to do it (anymore?). One dfx-command does the trick just as well.
- It links to Kyle's _personal_ website. While he's on the SDK team, the article is not an official (and therefore maintained) resource. By now it contains some outdated info (e.g. the URL to the SDK download), and the NNS interface shown in the screenshots is not accurate anymore.
- The link was combined with a bunch of deviations from the instructions. This makes it really hard to follow because people will have to switch between the two pages all the time.

Why I think the new link is fine:
- It's an official (and therefore maintained) resource.
- If we didn't link there, we'd be duplicating information, which makes maintenance much harder.
- It's literally where we send _everyone_ to start. We'd like to use at least that as common base knowledge and assume this in (almost?) all other examples.
- The requirements section is a _lot_ smaller this way.
- It's more clear that this is not a hard requirement, but only for live deployment. (Yes, this is only tangentially related to the link and more about how it's presented, but wanted to mention it anyways)

I hope that you can accept the new link with this reasoning.

(the question about which canister to deploy to is answered in a different comment)

    

-----
## Getting Started

- Initalize the project declarations and dfx processes:
    ```Bash
    # Deploys your canisters to the replica and generates your candid interface
    npm run init:dfx
    ```
- Serve the project locally:

    ```bash
    # Starts the replica, running in the background
    # Serves Angular and dfx canisters locally with hot reloading
    npm run start
    ```
Once the job completes, your application will be available at `http://localhost:4200`.

If you make any changes while in development mode, the project will recompile on both the Angular and Motoko side.

-----

## Deploying the Project to the IC:
- From the project root, deploy the project to your canisters provided in the [Prerequesites](#prerequesites).

    ```bash
    npm run deploy:dfx
    ```

-----

## Testing Locally:
- From the project root.

    ```bash
    npm run test
    ```

> -----

Reviewer: Can you also mention the normal dfx commands for this?

Author: Yes, can we make do with this for now as it is inferred in the actual command(s)?

Author: @sesi200

Reviewer: Can you at least point people to where they can find out what happens behind the scenes?

Author: I think I just need to rework the related scripts to work with `dfx delpoy` instead of `npm run deploy:dfx`

# hello-angular-motoko

> This is an example of locally implementing and deploying the Internet Computer's motoko project (0.10.1) along side Angular 14.

Reviewer: ```suggestion
This is an example of a Motoko project with an Angular 14 frontend.
```

-----------
## Getting started

> - [Install DFX](https://sdk.dfinity.org/docs/quickstart/local-quickstart.html). Please keep in mind the dfx cli currently only runs on Linux and Apple based PCs.

Reviewer: ```suggestion
- [Install DFX](https://internetcomputer.org/docs/current/developer-docs/quickstart/local-quickstart).
```
Updating with the current link, removing system requirements because that is a future stale bit of information.

- [Install Angular Globally](https://angular.io/guide/setup-local)
- Install npm packages from the project root:

    ```
    npm install
    ```
- Initalize the project declarations and dfx processes:
    ```Bash
    # Deploys your canisters to the replica and generates your candid interface
    npm run init:dfx
    ```

-----

## Serving the project locally
- You can use the following command:

    ```bash
    # Starts the replica, running in the background
    # Serves Angular and dfx canisters locally with hot reloading
    npm run start
    ```
    Once the job completes, your application will be available at `http://localhost:4200`.

>     If you make any changes while in development mode, the project will recompile on both the Angular and motoko side.

Reviewer: ```suggestion
    If you make any changes while in development mode, the project will recompile on both the Angular and Motoko side.
```

|`deploy:dfx`         |Deploy to IC           
|`restart:dfx`         |Restarts the local canister execution environment and web server processes. You can use this to Start the IC services as well.
|`restart:clean:dfx`|Starts the local canister execution environment and web server processes in a clean state by removing checkpoints from your project cache. You can use this to set your project cache to a new state when troubleshooting or debugging.|
|`generate:motoko`| Generate canister type declarations for motoko supported programming languages. This will generate the `src/declarations` folder and remove conflicts. These generated js files allow you to consume the motoko language service on the front end. i.e `const  ic_service = require('src/declarations/motoko').motoko;`

## Notes
- The npm `build`, `prebuild` and `prestart` commands are used by the IC processes. Using them may have unintended side effects with other build processes.
- Currently, the Typescript generated by DFX has trailing commas, which produces a conflict with Angular's default build configurations. Fixing this is beyound the scope of this example.

> - This example is intended to help those already familar with Angular to get started with the IC. This example is designed to be as straight forward as possible but it is not intended as a tool for learning Angular.

Reviewer: This should not link out to parts of external tutorials. The readme should be self-contained. Can you please simplify this a bit?

Side note: What does not work with `dfx deploy --network ic`? All projects we have so far work with a simple `dfx deploy`.

Author: Updated with self-contained Prerequesites. 
Side Note - `dfx deploy --network ic --no-wallet` deploys to the IC (https://ic0.app).  Without it, it deploys locally.

-----

### To learn more about working with dfx, see the following documentation available online:

- [Quick Start](https://sdk.dfinity.org/docs/quickstart/quickstart-intro.html)
- [SDK Developer Tools](https://sdk.dfinity.org/docs/developers-guide/sdk-guide.html)
- [Motoko Programming Language Guide](https://sdk.dfinity.org/docs/language-guide/motoko.html)
- [Motoko Language Quick Reference](https://sdk.dfinity.org/docs/language-guide/language-manual.html)

> - [JavaScript API Reference](https://erxue-5aaaa-aaaab-qaagq-cai.raw.ic0.app)

Reviewer: Most of this is not dfx-specific (as line 75 notes).

Author: Fixed.
